x/delayedack/ibc_middleware.go

```diff
@@ -1,323 +1,214 @@
-	}
+func SkipContext(ctx sdk.Context) sdk.Context {
+	return ctx.WithValue(ctxKeySkip{}, true)
+}
 
-	if rollappID == "" {
-		logger.Debug("Skipping IBC transfer OnRecvPacket for non-rollapp chain")
-		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
-	}
+// skip returns if the context contains the skip directive
+// Not intended to be used outside of module
+func skip(ctx sdk.Context) bool {
+	val, ok := ctx.Value(ctxKeySkip{}).(bool)
+	return ok && val
+}
 
-	err = im.ValidateRollappId(ctx, rollappID, rollappPortOnHub, rollappChannelOnHub)
-	if err != nil {
-		logger.Error("Failed to validate rollappID", "rollappID", rollappID, "err", err)
-		return channeltypes.NewErrorAcknowledgement(err)
-	}
+// OnRecvPacket handles the receipt of a packet and puts it into a pending queue
+// until its state is finalized
+func (w IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	l := w.logger(ctx, packet, "OnRecvPacket")
 
-	proofHeight, err := im.GetProofHeight(ctx, commontypes.RollappPacket_ON_RECV, rollappPortOnHub, rollappChannelOnHub, packet.Sequence)
-	if err != nil {
-		logger.Error("Failed to get proof height from packet", "err", err)
-		return channeltypes.NewErrorAcknowledgement(err)
+	if !w.Keeper.IsRollappsEnabled(ctx) {
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
 	}
 
-	finalized, err := im.CheckIfFinalized(ctx, rollappID, proofHeight)
-	if err != nil {
-		logger.Error("Failed to check if packet is finalized", "err", err)
-		return channeltypes.NewErrorAcknowledgement(err)
+	if skip(ctx) {
+		l.Info("Skipping because of skip delay ctx.")
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
 	}
 
-	if finalized {
-		logger.Debug("Skipping eIBC transfer OnRecvPacket as the packet proof height is already finalized")
-		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	transfer, err := w.GetValidTransferWithFinalizationInfo(ctx, packet, commontypes.RollappPacket_ON_RECV)
+	if err != nil {
+		l.Error("Get valid rollapp and transfer.", "err", err)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(err, "delayed ack: get valid transfer with finalization info"))
 	}
 
-	// Save the packet data to the store for later processing
-	rollappPacket := commontypes.RollappPacket{
-		RollappId:   rollappID,
-		Packet:      &packet,
-		Status:      commontypes.Status_PENDING,
-		Relayer:     relayer,
-		ProofHeight: proofHeight,
-		Type:        commontypes.RollappPacket_ON_RECV,
+	if !transfer.IsRollapp() || transfer.Finalized {
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
 	}
-	im.SetRollappPacket(ctx, rollappPacket)
 
-	logger.Debug("Set rollapp packet",
-		"rollappID", rollappPacket.RollappId,
-		"proofHeight", rollappPacket.ProofHeight,
-		"type", rollappPacket.Type)
+	rollappPacket := w.getSavedPacket(ctx, l, packet, transfer, relayer, commontypes.RollappPacket_ON_RECV, nil)
```

Comment:
it's passed once, not a big deal imo
prefer to avoid adding state where possible
(I agree, if it was passed several times or several layers deep then better to add state)